Initial upgrade to WebVR 1.1

Squareys opened the pull request:
**Hi everybody!**

This is the pullrequest for my work on #5028. Probably also closes #4309 and partially #4589.

I am already using this in an [example](https://github.com/mosra/magnum-examples/pull/33) for the [Magnum C++11/C++14 and OpenGL graphics engine](https://github.com/mosra/magnum) -- if you are interested, thanks to @mosra, you can view an online [preview version](http://magnum.graphics/showcase/webvr/) (requires WebGL 2).

As discussed in #5028, WebVR has changes quite a bit since the initial emscripten-vr API was written and therefore this is basically a big rewrite. I wrapped all essential features plus a couple of others which may come in handy. This does not wrap *all* WebVR functionality.

Looking forward to your reviews!

**Cheers, Jonathan.**

PS:

@kripken or @juj : One of you may want to cherry-pick c8aca03ba81a556a975fcd7a05042e04c7123ff2 right away, it fixes a minor typo in the firefox config for testing.

@promethe42 and @Noxalus: Any feedback on the API is welcome! I would also appreciate you trying the above example with your hardware: there seems to be an issue where the switch to VR (after requestPresent succeeds) seems to take quite a while on Oculus+Firefox while with [Threejs examples](https://threejs.org/examples/?q=webvr#webvr_cubes) it does not.

### Current State

All done, some smaller API questions to discuss (see below).

### Some Random Notes

 * All structures are named `VR*` to match the WebVR structure names.
 * Rather than "device", "display" is used to refer to a HMD to match the WebVR specification.
 * The "main loop" of WebVR displays is named *render loop*
 * All functions that take a `VRDisplayHandle` should return `int`, as they return `0` if the handle (index into an array) is invalid.

bjornbytes commented:
This is exciting, nice work!  If it's helpful, I've got a [branch](https://github.com/kripken/emscripten/compare/incoming...bjornbytes:lovr) laying around that implements some pieces of the WebVR 1.1 API.  Some stuff is messy and specific to my engine ([LÖVR](https://github.com/bjornbytes/lovr)), but it might still be useful.

Squareys commented:
While finishing the documentation, I noticed an inconsistency in the structures:
`VRVector3` and `VRQuaternion` have own structures (while in WebVR they are just float32 arrays), while the matrices in the `VRPose` structure are arrays.

I would probably prefer to make all of these arrays rather than types consistently throughout the API. Does anyone have an oppinion or preference?

kripken commented on src/library_vr.js:
```diff
@@ -1,237 +1,374 @@
+        display.requestAnimationFrame(display.mainLoop.runner);
+      },
+      runner: function() {
+        if(ABORT) return;
```
space after `if`

kripken commented on src/library_vr.js:
```diff
@@ -1,237 +1,374 @@
+        sourceStrPtr = {{{ makeGetValue('layerInitPtr', C_STRUCTS.VRLayerInit.source, 'void*') }}};
+
+        var source = null;
+        if(sourceStrPtr == 0) {
```
space after `if`

kripken commented on src/library_vr.js:
```diff
@@ -1,237 +1,374 @@
+
+    layerInit = new Array(layerCount);
+    for (var i = 0; i < layerCount; ++i) {
+        sourceStrPtr = {{{ makeGetValue('layerInitPtr', C_STRUCTS.VRLayerInit.source, 'void*') }}};
```
indentation here and elsewhere should be 2 spaces

Squareys commented:
@juj Any comments on the API design or documentation? :)

Squareys commented:
@kripken @juj Is there anything left to do here?

kripken commented:
Looks generally ok to me in terms of code style and so forth. Let's merge this in, and maybe a more in-depth look at the API can be done later when other people have time.

Squareys commented:
Awesome, thanks! 🎉

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+Initialization
+==============
+
+.. c:function:: int emscripten_vr_init()
```
I'll put in some review comments here, even though I do see this was already merged (which is fine) - I am very late to the review party here.

We would ideally favor returning an `EMSCRIPTEN_RESULT` rather than an `int` in these types of success/fail scenarios.

Squareys replied:
`EMSCRIPTEN_RESULT` is currently part of the html5 API.

While on the one hand, the common use of the vr API will require using the html5 API anyway (for a user-gesture callback to get permission to present), I don't know how much of a good idea is to have vr.h include html5.h.
There are some WebVR specific events like `"onvrdisplayconnect"` or `"onvrdisplayblur"`, which are currently not implemented, but later could or should using the html5 API?

I could also extract the emscripten result definitions into a separate header.
What do you think?

Squareys replied:
@juj I would highly appreciate your input on this to be able to wrap up #5640 
I saw the WebVR API has been released in the meanwhile, which makes me a little uncomfortable with breaking the return values in this way now...

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+.. c:function:: int emscripten_vr_ready()
+
+	Check whether the VR API has finished initializing VR displays.
```
Yikes, polling "is init done?" API - we'd want a waitable handle here, which a token that one can likewise poll, wait or get a "when it's done" callback for. Though full support for that will need multithreading enabled.

Squareys replied:
Yeah, I didn't really like this part either...

> Though full support for that will need multithreading enabled.

You are currently working on this, right? For now I could just add a callback to the `emscripten_vr_init()` function.

juj replied:
Yeah, that's right - that's why it's fresh on my mind right now, though it's not quite landed just yet to be usable, so it will be a moment until that gets in to be used by other APIs like this, but I'll note when that is.

Squareys replied:
I added a callback for now. Should I add a TODO comment or similar for making this a waitable handle?

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+.. c:function:: int emscripten_vr_get_frame_data(VRDisplayHandle handle)
+
+	Get view matrix, projection matrix, timestamp and head pose for current frame.
+	Only valid when called from within a render loop callback.
```
Meaning of return value is not documented here - is it a 0/1 success boolean?

juj replied:
Also, based on documentation alone, for someone not aware about the internals, when this says that it gets the matrices, one can ask "where does it get them to?" since it does not return matrix values in this function signature, so that would be good to be documented as well.

Squareys replied:
Right, will fix that!

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+	.. c:member:: VRVector3 position
+
+		Position, valid only if ``poseFlags & VR_POSE_POSITION == 0``.
```
Should this not be the other way around? In above it says `which has a bit set for every valid value`, but here it is testing the other way around, i.e. one'd expect

```
, valid only if ``(poseFlags & VR_POSE_POSITION) != 0``.
```

(also I think it will be good to have the extra parentheses here on the docs as well, since otherwise LLVM will issue a compiler warning if one copy pastes the example code as-is)

Squareys replied:
> Should this not be the other way around?

Yes, definitely, sorry about that!

Squareys commented:
@juj Thanks for the review! I will open up a pullrequest to address your comments until end of the week.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+	Minor version of the WebVR API currently supported by the browser.
+
+	:returns: minor version of WebVR, or `-1` if not supported or API not initialized.
```
Here I was looking what would happen if one deinitializes the API, and if the major and minor version numbers would return -1 after deinitialization, but then noticed that deinitialization paths are not supported at all. There's a couple of reasons why we will at some point need a graceful deinit path as well:

- some pages may want to host multiple apps, or run a specific app multiple times, and they may want to depend on deinitializing the previous execution after being done, to make room for the next run,
- some apps might only run VR in certain cases, and toggle VR on/off multiple times during run,
- multithreaded apps - we'll need to define how vr initialization works with respect to multiple threads. Will emscripten_vr_init() be process-wide? or thread-specific? If thread-specific, can only one thread at a time have VR initialized, or can multiple threads initialize, and then each will have their own instance running?
- once WebVR 2.0 lands, we may need to be able to initialize both 1.1 or 2.0 version (not sure how this will pan out)

Just throwing these kind of questions on the table, for further reference and ideas, I don't expect we'll need to rush into any additions here.

Squareys replied:
I need to implement `emscripten_vr_deinit()` for freeing the displayNames anyway. So, how would you prefer the version being handled? Set it back to -1/-1?

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+	:returns: handle for a VR display.
+	:rtype: VRDisplayHandle
+
+.. c:function:: char* emscripten_vr_get_display_name(VRDisplayHandle handle)
```
The return value here might be good to be a `const char *`. (though `const` will need to be used in an old C aware way, which did not have `const` until C99)

juj commented on src/library_vr.js:
```diff
@@ -1,237 +1,373 @@
+    buffer = _malloc(len + 1);
+    stringToUTF8(displayName, buffer, len + 1);
+
+    return buffer;
```
It looks like we have a memory leak here, the above code `malloc`s, and responsibility to `free` is assigned to the caller, but the documentation does not state that. I think rather than assigning burden to `free` to caller, it will be better to statically allocate on first call, and hold on to the allocation, returning the same allocated pointer each time, and WebVR subsystem deinit will then eventually `free` the allocation.

Or alternatively, user passes in the memory block where to write to, but that is probably unneedlessly complicated, so the current signature might be preferred, since these are expected to be static.

Squareys replied:
I like the static string, as this is how `glGetString()` does things. I also added `emscripten_vr_deinit()` along with it.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+	:param int displayIndex: index of display (inclusive 0 to exclusive :c:func:`emscripten_vr_count_displays`).
+	:returns: handle for a VR display.
+	:rtype: VRDisplayHandle
```
Here we'll need to be precise what happens if one connects and disconnects displays in different orders, e.g. connect display 1, connect display 2, disconnect display 1, disconnect display 2. Are the `VRDisplayHandle` objects the unique identifiers that are not affected by connects and disconnects, and on connect and disconnect, only the `emscripten_vr_count_displays()` and `emscripten_vr_get_display_handle()` functions are affected?

If one connects a vr display, then disconnects it, and then reconnects the same display, will the `VRDisplayHandle` be the same for that display, or can it be different?

Squareys replied:
The displays will always be present and stay at the same location (in the array), but have their `emscripten_vr_display_connected()` be `true/false` accordingly.
I am usure if a entirely new display is even recognized after `emscripten_vr_init()` at all. I think this would require handling the `"ondisplayconnected"` event.
Will do some research to clarify...

Squareys replied:
`"ondisplayconnected"` passes the connected display to the event handler. This could obviously be a display that already exists in the `WebVR.displays` array. But we currently don't handle this event at all, displays need to therefore be connected before the page is loaded.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+	:returns: name of the VR display or `0 (NULL)` if the handle is invalid.
+	:rtype: char*
+
+.. c:function:: bool emscripten_vr_get_display_connected(VRDisplayHandle handle)
```
The documentation here refers to `emscripten_vr_get_display_connected`, but it looks like the implementation is for a function `emscripten_vr_display_connected` instead.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+		the handle is invalid.
+	:rtype: bool
+
+.. c:function:: bool emscripten_vr_get_display_presenting(VRDisplayHandle handle)
```
Also here `emscripten_vr_get_display_presenting` vs `emscripten_vr_display_presenting`.

juj replied:
Haven't jumped in to read the spec in detail on these nuances, though here I wonder if apps are expected to be polling these functions (e.g. in the beginning of each render frame?) for whether they are still presenting and connected, or are there notification events involved for disconnection and "lost present"? (or for the same way, "connected" and "acquired present"?)

Squareys replied:
There are events for this, but these two functions are implemented using a property of `VRDisplay`.

Most examples seem to check `display.isPresenting` every frame (e.g. [display present example from webvr.info](https://github.com/toji/webvr.info/blob/master/samples/03-vr-presentation.html#L204)).

As noted in [this comment](https://github.com/kripken/emscripten/pull/5508#discussion_r143470069), WebVR specific events are not implemented with this pullrequest currently.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+	:param VREye whichEye: which eye to query parameters for.
+	:param VREyeParameters eyeParam: receives the parameters for requested eye.
+	:returns: |display-function-return-doc|
+	:rtype: bool
```
On all of these success/fail functions, I'd like to see `EMSCRIPTEN_RESULT` being used, though I do notice it would flip the success/fail meaning the other way around and that might confuse, so your judgement call. The one future proofing scenario that I'm thinking here where EMSCRIPTEN_RESULT might be useful is to carry minimal amount of error reasons, e.g. `EMSCRIPTEN_RESULT_NOT_SUPPORTED` if browser does not support WebVR, and `EMSCRIPTEN_RESULT_INVALID_TARGET` if e.g. one lost access to the VR device, or `EMSCRIPTEN_RESULT_INVALID_PARAM` if user passed in something bad, and so on. Also  `EMSCRIPTEN_RESULT_DEFERRED` and/or `EMSCRIPTEN_RESULT_FAILED_NOT_DEFERRED` may be useful when making the API multithreading capable, in the scenarios where we may need to proxy stuff over to the main thread.

Squareys replied:
I agree and I do not shy the effort required to change all the return values. Please see my comment [here](https://github.com/kripken/emscripten/pull/5508#discussion_r143470069) about how to handle the dependency on html5.h.

> though I do notice it would flip the success/fail meaning

Then it is important to do this now before this gets released (and I'm probably the only one who needs to adapt his code)! 👍

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+loop can be set per display and will act like a main loop with timing mode ``EM_TIMING_RAF`` until the
+display is requested to present, as of which it will run at the VR display's refresh rate.
+
+.. c:function:: void emscripten_vr_set_display_render_loop(VRDisplayHandle handle, em_vr_callback_func callback)
```
Here there's also a question about what is the thread context where the render loop is called - probably we want to have the same as the multithreaded emscripten set main loop, which runs the render loop in the context of the calling thread. Then there's a question of what if a vr render loop is registered multiple times in one thread for one device, and what happens if a multiple vr render loops for the same device are registered in separate threads.

Squareys replied:
> probably we want to have the same as the multithreaded emscripten set main loop

I have not used multithreading in Javascript before, so I don't know the consequences of one or the other. If calling WebGL is only allowed from the main loop, then that is obviously what we want, yes :)

> what if a vr render loop is registered multiple times in one thread for one device

It is asserted that there can only be one renderloop per display. The following `.. tip::` section clarifies that there can only be one. Or what do you mean?

> multiple vr render loops for the same device are registered in separate threads

I would expect the memory of WebVR module to be shared between the devices and therefore the same assertion would fail if a render loop is already set.

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+.. tip:: There can be only *one* render loop function per VR display. To change the render loop function, first :c:func:`cancel <emscripten_vr_cancel_display_render_loop>` the current loop, and then call this function to set another.
+
+.. c:function:: void emscripten_vr_set_display_render_loop_arg(VRDisplayHandle handle, em_vr_callback_func callback, void* arg)
```
Owch, here we followed the same paradigm as the original set main loop. That is great, though the multiple overloads there have been kind of historical, so we want to redesign this at some point.

Squareys replied:
Is there already a new design? I'd be happy to simply remove the `emscripten_set_display_render_loop` function (without the userData).

juj commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+
+	Submit the current state of canvases passed via VRLayerInit to
+	:c:func:`emscripten_vr_request_present` to be rendered to the VR display.
+	Only valid when called from within a render loop callback.
```
This is very much like regular GL/D3D present? (https://www.khronos.org/registry/EGL/sdk/docs/man/html/eglSwapBuffers.xhtml, https://msdn.microsoft.com/en-us/library/windows/desktop/bb174576(v=vs.85).aspx) I.e. until one calls `submitFrame`, nothing is actually shown on screen?

What happens if one returns from the VR display callback without calling `emscripten_vr_submit_frame` function? I suppose the previous frame that was presented with an earlier `emscripten_vr_submit_frame` call will continue to show on the screen?

Does one need to always pair each VR display callback with exactly one `emscripten_vr_submit_frame` call? What happens if one calls that function twice in a callback? Or if one called that exactly once every second callback?

Squareys replied:
>  I suppose the previous frame that was presented with an earlier emscripten_vr_submit_frame call will continue to show on the screen

The VR driver will probably handle this as a missed a frame and try to timewarp a new frame according to the head pose. If you do that a couple times in a row, it will either fade out of your application to some driver specific scene (which notifies the user that the app is not cooperating) or maybe just display the last frame. The spec doesn't seem to handle this.

> Does one need to always pair each VR display callback with exactly one emscripten_vr_submit_frame call?

Rather: pair every `emscripten_vr_get_frame_data` with `emscripten_vr_submit_frame`... the draft for 2.0 already changes how this works.

Is that a question out of interest or should this be clarified in the docs? In the latter case: I am unsure whether (re-)explaining the WebVR API in the emscripten docs makes sense.

Squareys commented on site/source/docs/api_reference/vr.h.rst:
```diff
@@ -0,0 +1,378 @@
+	If the request is successful `callback` will be called with `userData` and the render
+	loop will continue rendering at the refresh rate of the VR display.
+
+	Must be called from a user callback (see :ref:`HTML5 API <html5-h>`).
```
How would I check if I'm in a callback? There is [JSEvents.inEventHandler](https://github.com/kripken/emscripten/blob/master/src/library_html5.js#L126) and probably `currentEventHandler`. Can I just depend on JSEvents like that?

Squareys commented:
@juj I addressed some of the comments with #5640.

While researching to answer your questions I found that WebVR 2.0 spec draft uses "device" rather than "display" to refer to HMDs again 😅 
Alot of things will change with that again and the 1.1 spec is marked as [pre-release](https://w3c.github.io/webvr/), so I am hesitant to put too much more effort into WebVR 1.1 support.

Definitely the return values (with #5640, once we clarified how to handle the dependency on html5.h) and *maybe* the events in a couple of months, but that depends on whether the WebVR 2.0 spec or something else actually stable gets released in the meantime.
